File: Difficulty: Hard/Burning Tree/burning-tree.cpp
//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    Node *left;
    Node *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};


Node *buildTree(string str) {
    // Corner Case
    if (str.length() == 0 || str[0] == 'N')
        return NULL;

    // Creating vector of strings from input
    // string after spliting by space
    vector<string> ip;

    istringstream iss(str);
    for (string str; iss >> str;)
        ip.push_back(str);

    // Create the root of the tree
    Node *root = new Node(stoi(ip[0]));

    // Push the root to the queue
    queue<Node *> queue;
    queue.push(root);

    // Starting from the second element
    int i = 1;
    while (!queue.empty() && i < ip.size()) {

        // Get and remove the front of the queue
        Node *currNode = queue.front();
        queue.pop();

        // Get the current Node's value from the string
        string currVal = ip[i];

        // If the left child is not null
        if (currVal != "N") {

            // Create the left child for the current Node
            currNode->left = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->left);
        }

        // For the right child
        i++;
        if (i >= ip.size())
            break;
        currVal = ip[i];

        // If the right child is not null
        if (currVal != "N") {

            // Create the right child for the current Node
            currNode->right = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->right);
        }
        i++;
    }

    return root;
}


// } Driver Code Ends
//User function Template for C++

/*
struct Node {
    int data;
    Node *left;
    Node *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};
*/
class Solution {
  public:
  
    Node* find_parent(Node* root,int target,unordered_map<Node*,Node*>&parent_tracker){
        queue<Node*>q;
        q.push(root);
         Node*res=NULL;
        while(!q.empty()){
            Node*curr=q.front();
            if(curr->data==target){
                res=curr;
            }
            q.pop();
            if(curr->left){
                q.push(curr->left);
                parent_tracker[curr->left]=curr;
            }
            if(curr->right){
                q.push(curr->right);
                parent_tracker[curr->right]=curr;
            }
        }
       return res; 
    }
    int minTime(Node* root, int target) 
    {
       unordered_map<Node*,Node*>parent_tracker;
       Node*res= find_parent(root,target,parent_tracker);
       unordered_map<Node*,bool>visited;
       queue<Node*>q;
       q.push(res);
       visited[res]=true;
    
       int time=0;
       while(!q.empty()){
            bool check=false;
           int size=q.size();
           for(int i=0;i<size;i++){
               Node* node=q.front();
               q.pop();
               if(node->left && visited[node->left]==false){
                   q.push(node->left);
                   visited[node->left]=true;
                   check=true;
               }
                 if(node->right && visited[node->right]==false){
                   q.push(node->right);
                   visited[node->right]=true;
                   check=true;
               }
               if(parent_tracker[node] && visited[parent_tracker[node]]==false){
                  visited[parent_tracker[node]]=true;
                   q.push(parent_tracker[node]);
                   check=true;
               }
            }
            if(check==true)time++;
       }
       return time;
     }
};



//{ Driver Code Starts.

int main() 
{
    int tc;
    scanf("%d ", &tc);
    while (tc--) 
    {    
        string treeString;
        getline(cin, treeString);
        // cout<<treeString<<"\n";
        int target;
        cin>>target;
        // cout<<target<<"\n";

        Node *root = buildTree(treeString);
        Solution obj;
        cout<<obj.minTime(root, target)<<"\n"; 

        cin.ignore();

    }


    return 0;
}

// } Driver Code Ends

File: Difficulty: Hard/Burning Tree/burning-tree.py
from typing import Tuple

class Solution:
    def _depth(self, root: 'Node | None') -> int:
        if root is None:
            return 0
        return 1 + max(self._depth(root.left), self._depth(root.right))

    def burn(self, root: 'Node | None', target: int) -> Tuple['int | None', int]:
        if root is None:
            return None, 0
        if target == root.data:
            return self._depth(root) - 1, 0
        left_elapsed, left_depth = self.burn(root.left, target)
        right_elapsed, right_depth = self.burn(root.right, target)
        if left_elapsed is None and right_elapsed is None:
            return None, 1 + max(left_depth, right_depth)
        if left_elapsed is None:
            assert right_elapsed is not None
            right_depth += 1
            return max(right_elapsed, left_depth + right_depth), right_depth
        left_depth += 1
        return max(left_elapsed, left_depth + right_depth), left_depth
        
    def minTime(self, root, target):
        # code here
        return self.burn(root, target)[0] or 0



#{ 
 # Driver Code Starts
#Initial Template for Python 3

from collections import deque


# Tree Node
class Node:

    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None


# Function to Build Tree
def buildTree(s):
    #Corner Case
    if (len(s) == 0 or s[0] == "N"):
        return None

    # Creating list of strings from input
    # string after spliting by space
    ip = list(map(str, s.split()))

    # Create the root of the tree
    root = Node(int(ip[0]))
    size = 0
    q = deque()

    # Push the root to the queue
    q.append(root)
    size = size + 1

    # Starting from the second element
    i = 1
    while (size > 0 and i < len(ip)):
        # Get and remove the front of the queue
        currNode = q[0]
        q.popleft()
        size = size - 1

        # Get the current node's value from the string
        currVal = ip[i]

        # If the left child is not null
        if (currVal != "N"):

            # Create the left child for the current node
            currNode.left = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.left)
            size = size + 1
        # For the right child
        i = i + 1
        if (i >= len(ip)):
            break
        currVal = ip[i]

        # If the right child is not null
        if (currVal != "N"):

            # Create the right child for the current node
            currNode.right = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.right)
            size = size + 1
        i = i + 1
    return root


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        line = input()
        target = int(input())
        root = buildTree(line)
        print(Solution().minTime(root, target))

        print("~")

# } Driver Code Ends